File: frontend/src/utils/priceActionTaxonomy.ts
import type { EntryTrigger, LocationTag, MarketState, Trade, TradeThesis } from "../types";
import taxonomy from "../../../shared/price_action_taxonomy.json";

export interface BilingualTaxonomyItem<T extends string> {
  value: T; english: string; chinese: string; englishDescription?: string; chineseDescription?: string;
}

interface ContractItem {
  value: string;
  english: string;
  chinese: string;
  order: number;
  english_description?: string;
  chinese_description?: string;
}

function contractItems<T extends string>(items: ContractItem[]) {
  return [...items]
    .sort((left, right) => left.order - right.order)
    .map((item) => ({
      value: item.value as T,
      english: item.english,
      chinese: item.chinese,
      ...(item.english_description ? {
        englishDescription: item.english_description,
        chineseDescription: item.chinese_description,
      } : {}),
    })) as BilingualTaxonomyItem<T>[];
}

export const marketStateOptions = contractItems<MarketState>(taxonomy.market_state);
export const tradeThesisOptions = contractItems<TradeThesis>(taxonomy.trade_thesis);
export const entryTriggerOptions = contractItems<EntryTrigger>(taxonomy.entry_trigger);
export const locationTagOptions = contractItems<LocationTag>(taxonomy.location_tag);

export function findTaxonomyItem<T extends string>(items: BilingualTaxonomyItem<T>[], value: T | null | undefined) {
  return items.find((item) => item.value === value) ?? null;
}

const legacyState: Record<string, MarketState> = { strong_trend: "strong_trend", narrow_channel: "narrow_channel", weak_trend: "broad_channel", broad_channel: "broad_channel", trading_range: "trading_range", breakout_mode: "breakout_mode", uncertain: "unclear", opening_range: "unclear", gap_open: "unclear" };
const legacyThesis: Record<string, TradeThesis> = { breakout: "breakout", pullback: "pullback_continuation", failed_breakout: "failed_breakout", reversal: "major_reversal", left_side_bottom_pick: "major_reversal", early_reversal: "major_reversal", bottom_pick: "major_reversal", h1_h2_l1_l2: "other", wedge: "other", double_top_bottom: "other", inside_bar_triangle: "other", opening_range: "other", gap_open: "other", other: "other" };
const legacyTrigger: Record<string, EntryTrigger> = { h1_h2_l1_l2: "h1_h2_l1_l2", wedge: "wedge", double_top_bottom: "double_top_bottom", inside_bar_triangle: "inside_bar_triangle" };

export function displayClassification(trade: Pick<Trade, "market_state" | "trade_thesis" | "entry_trigger" | "location_tags" | "setup" | "market_context">) {
  const legacy = !trade.market_state || !trade.trade_thesis || !trade.entry_trigger;
  return {
    legacy,
    marketState: trade.market_state ?? legacyState[trade.market_context] ?? null,
    tradeThesis: trade.trade_thesis ?? legacyThesis[trade.setup] ?? null,
    entryTrigger: trade.entry_trigger ?? legacyTrigger[trade.setup] ?? (trade.setup ? "other" : null),
    locationTags: trade.location_tags ?? [],
  };
}

export function legacyMirrors(state: MarketState, thesis: TradeThesis) {
  const setup: Record<TradeThesis, string> = { pullback_continuation: "pullback", breakout: "breakout", breakout_pullback: "breakout", failed_breakout: "failed_breakout", range_reversal: "reversal", major_reversal: "reversal", other: "other" };
  const context: Record<MarketState, string> = { strong_trend: "strong_trend", narrow_channel: "narrow_channel", broad_channel: "broad_channel", trading_range: "trading_range", breakout_mode: "breakout_mode", unclear: "uncertain" };
  return { setup: setup[thesis], market_context: context[state] };
}

File: backend/app/services/price_action_taxonomy.py
"""Canonical price-action contract and legacy compatibility mapping."""

from __future__ import annotations

import json
from pathlib import Path

TAXONOMY_PATH = Path(__file__).parents[3] / "shared" / "price_action_taxonomy.json"
with TAXONOMY_PATH.open(encoding="utf-8") as taxonomy_file:
    TAXONOMY_CONTRACT = json.load(taxonomy_file)

MARKET_STATES = tuple(item["value"] for item in TAXONOMY_CONTRACT["market_state"])
TRADE_THESES = tuple(item["value"] for item in TAXONOMY_CONTRACT["trade_thesis"])
ENTRY_TRIGGERS = tuple(item["value"] for item in TAXONOMY_CONTRACT["entry_trigger"])
LOCATION_TAGS = tuple(item["value"] for item in TAXONOMY_CONTRACT["location_tag"])

CONTEXT_TO_STATE = {
    "strong_trend": "strong_trend", "narrow_channel": "narrow_channel",
    "weak_trend": "broad_channel", "broad_channel": "broad_channel",
    "trading_range": "trading_range", "breakout_mode": "breakout_mode",
    "uncertain": "unclear", "opening_range": "unclear", "gap_open": "unclear",
}
SETUP_TO_CLASSIFICATION = {
    "breakout": ("breakout", "other"), "pullback": ("pullback_continuation", "other"),
    "failed_breakout": ("failed_breakout", "other"), "reversal": ("major_reversal", "other"),
    "left_side_bottom_pick": ("major_reversal", "other"), "early_reversal": ("major_reversal", "other"),
    "bottom_pick": ("major_reversal", "other"), "h1_h2_l1_l2": ("other", "h1_h2_l1_l2"),
    "wedge": ("other", "wedge"), "double_top_bottom": ("other", "double_top_bottom"),
    "inside_bar_triangle": ("other", "inside_bar_triangle"), "opening_range": ("other", "other"),
    "gap_open": ("other", "other"), "other": ("other", "other"),
}
THESIS_TO_SETUP = {
    "pullback_continuation": "pullback", "breakout": "breakout", "breakout_pullback": "breakout",
    "failed_breakout": "failed_breakout", "range_reversal": "reversal", "major_reversal": "reversal", "other": "other",
}
STATE_TO_CONTEXT = {
    "strong_trend": "strong_trend", "narrow_channel": "narrow_channel", "broad_channel": "broad_channel",
    "trading_range": "trading_range", "breakout_mode": "breakout_mode", "unclear": "uncertain",
}
UNCONFIRMED_SETUPS = {"left_side_bottom_pick", "early_reversal", "bottom_pick"}


def ordered_location_tags(values: list[str] | None) -> list[str]:
    selected = set(values or [])
    return [value for value in LOCATION_TAGS if value in selected]


def classification_from_legacy(setup: str | None, context: str | None) -> dict:
    thesis, trigger = SETUP_TO_CLASSIFICATION.get(setup or "", ("other", "other"))
    tags = [value for value in (context, setup) if value in {"opening_range", "gap_open"}]
    return {
        "market_state": CONTEXT_TO_STATE.get(context or "", "unclear"),
        "trade_thesis": thesis,
        "entry_trigger": trigger,
        "location_tags": ordered_location_tags(tags),
        "location_decision": "selected" if tags else None,
        "reversal_confirmation": "unconfirmed" if setup in UNCONFIRMED_SETUPS else None,
        "is_unconfirmed_reversal": setup in UNCONFIRMED_SETUPS,
    }


def add_legacy_mirrors(data: dict) -> dict:
    if data.get("trade_thesis"):
        mirrored = THESIS_TO_SETUP.get(data["trade_thesis"])
        if mirrored:
            data["setup"] = mirrored
    if data.get("market_state"):
        mirrored = STATE_TO_CONTEXT.get(data["market_state"])
        if mirrored:
            data["market_context"] = mirrored
    return data
